Added protection from performing illegal mathematical operations

Craxic opened the pull request:
These changes cause the division block to raise an exception when a division by 0 is attempted, rather than returning "NaN" or "Infinity" (this is very confusing for people new to programming). It also prevents the modulus operation block from performing modulus by 0, and causes the monadic operation block (sqrt, pow) from doing things like `sqrt(-1)`.

Finally, it prevents sprites from moving to Infinity or NaN locations.

Craxic committed:
Added protection from performing illegal mathematical operations

brianharvey commented:
Hmm, why do some of these check for infinity and others not (e.g., tan 90)?  Also, I don't think changing infinity or NaN to 0 in the motion blocks makes sense.  If anything, give an error message.  But the changes to the numeric operators should make it impossible to get infinity or NaN as an input to the motion blocks anyway, right?   And what's special about motion blocks?  I'd say it's more important to avoid an infinite REPEAT!

Craxic commented:
Hi Brian,

The ones which don't check for Infinity are the operator blocks. The reason I decided not to do this was because I thought that Infinity represents an overflow, rather than an illegal operation (like NaN represents). For example, e^100000000000000 yields Infinity despite it making logical sense. Infinity even behaves correctly in some blocks, like "<". However, while I was investigating this to see if I was correct about the above, I noticed that Infinity times 0 is actually NaN. So if we decide to leave them in, we'll have to put checks in the multiplication operator too.

Changes to the numeric operators should theoretically make it impossible to get such numbers into the movement blocks, but I wasn't confident enough that I had removed all possible ways to generate NaN, so I added them in as a last line of defence. Regardless of NaN, doing things like rotating to Infinity causes very strange behaviour without these checks in place.

And yep, avoiding an Infinity repeat is something we should do. My bad. I'll make the changes. But let me know how you feel about the above.

Thanks!

jmoenig commented:
Thanks! I'm willing to "bless" this PR and pull it. Brian, what do you say?

brianharvey commented:
I'm not sure.  If I'm understanding correctly, this means that 0/0 is considered an error, but 4/0 is a legitimate (infinite) value.  I can see how a set theorist might want that behavior, but for high school students, if we're catching either I think we should catch both.  (And we need dyadic atan, maybe in a math library.)

jmoenig commented:
Wait! the way I understand the code is that whenever the denominator is zero *or* the quotient is NaN it throws an error. That includes raising an error for the ```4 / 0``` case, right? Looks legit to me, I like this contribution.

I also agree about dyadic atan, yes we should have that, and yes, that could go into a library! (why don't you open an issue about it, it's a good little project?)

brianharvey commented:
Okay, if you say so.

Why don't I just write it instead of filing an issue?  :-)

cycomachead commented on threads.js:
```diff
@@ -2036,14 +2036,32 @@ Process.prototype.reportProduct = function (a, b) {
 };
 
+function QuotientNaNException(numerator, denominator) {
+    this.name = "Quotient exception";
```
Can we avoid using the word "exception"? I think "error" is much more clear to the average human.

brianharvey replied:
Yes.

cycomachead commented on threads.js:
```diff
@@ -2036,14 +2036,32 @@ Process.prototype.reportProduct = function (a, b) {
 
+function QuotientNaNException(numerator, denominator) {
+    this.name = "Quotient exception";
+    this.message = "illegal division operation: (" + numerator + ") / (" + denominator + ")";
```
Also, maybe something other than "illegal", too, though I'm somewhat less opposed to this. Perhaps "Unsupported"?

brianharvey replied:
Hmm I made a comment here but it didn't stick...
No to "Unsupported," because that sounds as if this should work but doesn't because of a Snap! deficiency.  "Undefined"?

cycomachead replied:
I like the word undefined. :+1:

cycomachead commented on threads.js:
```diff
@@ -2036,14 +2036,32 @@ Process.prototype.reportProduct = function (a, b) {
 
+function ModulusNaNException(numerator, denominator) {
+    this.name = "Modulus exception";
+    this.message = "illegal modulus operation: (" + numerator + ") / (" + denominator + ")";
```
Perhaps we could should this as `num % denom` or `num mod denom` The `/` symbol seems confusing to me...

brianharvey replied:
Yes to mod, no to %.
These are good catches -- thanks!

cycomachead commented:
I think these changes make sense overall, though I left comments about cleaning up the error messages. Currently the sound just a bit to "programmer"-y for my liking, and exceptions and errors are things that confuse students more than ever. I think we should make them as friendly as possible.

Few related things:
1. Should we maybe have a "reason" in the error message?
Some students may not realize that you can't divide by 0.

2. Why not have few global functions?
```
function MathException(name, operator, val1, val2, reason?) {
...
}
```
And reduce the repeition of all the other stuff.

3. We need a few calls to `localize()` around some of those messages, so that we can properly translate the strings.

* `X/0` I do agree with Brian. IMO this is more ZeroDivisionError than `Infinity`, but I could see the argument for leaving it as is.

* Can we do the same for all numeric primitives? It would be nice to at least bust `NaN` from a student's view.

> Also, I don't think changing infinity or NaN to 0 in the motion blocks makes sense. If anything, give an error message.

Hmm, personally, I don't know if an error is helpful. (Though I don't have any other than gut and intution for that reason, but it just feels like the Scratch philosophy of having reasonable defaults makes sense here.)

I just realized that we _should_ at do something for `Infinity` values because it does lead to awkward behavior.


Also tag #602

cycomachead commented on threads.js:
```diff
@@ -2036,14 +2036,32 @@ Process.prototype.reportProduct = function (a, b) {
-    return ((x % y) + y) % y;
+    var result = ((x % y) + y) % y;
+    if (isNaN(result)) {
+        throw new ModulusNaNException(+a, +b);
```
Minor: but probably better for those to just be `x` and `y`

cycomachead commented on objects.js:
```diff
@@ -3445,7 +3445,12 @@ SpriteMorph.prototype.setPosition = function (aPoint, justMe) {
-        dist = steps * this.parent.scale || 0;
+        dist;
+    
+    if (isNaN(steps) || !isFinite(steps)) {
```
I don't know how @jmoenig Feels about this, but I feel like this should be a shared function.

``` function defaultFiniteValue(num, default) {
     if (isNaN(steps) || !isFinite(steps)) {
        return default;
    } else {
        return value;
    }
}
```

Actually, you could probably write that with `?:`.

Then you can just do `steps = defaultValue(steps, 0)` which I think is much easier to read.

(Feel free to rename the function, but I do think it would clean things up.)

brianharvey commented:
About reasonable defaults:  If you give a MOVE command and nothing happens at all, your first thought isn't going to be that your incorrect expression defaulted to zero.  It'll be that somehow you didn't get to the MOVE block at all.  If we were going to do something Scratch-like, it'd be to move the sprite to the edge of the stage.  But I still vote for an error message.

jmoenig commented:
Let's stick with zero for unsupported MOVE arguments! @Craxic's point is that if we catch unsupported values at the operator level we're going to catch them earlier anyway. Scratch's way would not be any surprising action like moving to the edge of the stage (why?), it always is to simply noop.

Craxic committed:
Prevent infinite repeat

Craxic committed:
Consolidated errors into one error type

Craxic committed:
Added defaultFiniteValue function.

Craxic commented:
Hi guys,

I've tried to address all your concerns. Just a couple of things to note:

- I'm not sure what the convention is for formatting localized strings. For example, ```this.name = localize("%s operation error").replace("%s", localOp);```. Not sure if that's how I should be doing it.

- I agree with Brian that its probably better to give an error rather than just moving to the origin with the move blocks. Up to Jens.

Matt.

cycomachead commented:
Usually I think the convention is simply `str + ' ' + localize(constantStr)`, but Jens could advise. :)

jmoenig commented on objects.js:
```diff
@@ -141,6 +141,15 @@ var StagePrompterMorph;
 var Note;
 var SpriteHighlightMorph;
 
+function defaultFiniteValue(num, def) {
```
okay, can we move these functions from the global namespace into the constructors they work in? Here it would be SpriteMorph, with a shadow in StageMorph?

jmoenig commented on threads.js:
```diff
@@ -1629,11 +1629,21 @@ Process.prototype.doForever = function (body) {
     this.pushContext();
 };
 
+function DoRepeatError(value) {
```
same with these functions, can we put them into Process?

jmoenig commented on threads.js:
```diff
@@ -2024,26 +2034,53 @@ Process.prototype.reportTypeOf = function (thing) {
 
 // Process math primtives
 
+function UndefinedOperationError(opName, argA, symbol, argB) {
```
same here... Process?

jmoenig commented on threads.js:
```diff
@@ -2024,26 +2034,53 @@ Process.prototype.reportTypeOf = function (thing) {
+    }
+}
+
+function checkBinaryOperationResult(result, opName, argA, symbol, argB) {
```
and here... also in Process? Would that work?

Craxic committed:
Addressed Jens' concerns

Craxic commented:
Hi Jens,

I've rebased the changes and addressed your concerns. It should be ready to merge now, hopefully.

Matt.

jmoenig commented:
Thanks! I'll probably pull this with the next "bigger" release.
